// node-fest/lib/common/result.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Result = void 0;
var util = require('util');
var chalk = require('chalk');
var format = require('string-format');
var Result = /** @class */ (function () {
    function Result() {
        this.pass_count = 0;
        this.fail_count = 0;
        this.total_count = 0;
    }
    Result.prototype.markPass = function (tc) {
        this.pass_count++;
        if (tc) {
            console.log(chalk.green("[" + tc.line + "][PASS] TEST PASS :)"));
        }
    };
    Result.prototype.markFail = function (tc, msg) {
        this.fail_count++;
        if (tc && msg) {
            console.log(chalk.red("[" + tc.line + "][FAIL] Test case failed: " + JSON.stringify(tc) + ", msg: " + msg));
        }
    };
    Result.prototype.printResult = function () {
        var result = util.format("\n\n\
        =======================================================\n\
                                SUMMARY                        \n\
        =======================================================\n\
        Pass Count: %s\n\
        Fail Count: %s\n\
        Total TC: %s\n\
        Pass Percentage: %s\%\n\
        =======================================================\
        ", this.pass_count, this.fail_count, this.pass_count + this.fail_count, (this.pass_count * 100 / (this.pass_count + this.fail_count)));
        if (this.fail_count == 0) {
            console.log(chalk.green(result));
        }
        else {
            console.log(chalk.red(result));
        }
    };
    return Result;
}());
exports.Result = Result;
